aten/src/ATen/core/ivalue_inl.h

```diff
@@ -762,30 +732,62 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
           " and ",
           devices[idx]);
     }
-    return c10::impl::getDeviceGuardImpl(deviceType);
+    return deviceType;
   }
 
-  // We need devices to be sorted in order to use set_difference.
-  static std::vector<c10::DeviceIndex> getSortedIndicesOfDevices(
-      const c10::impl::DeviceGuardImplInterface* impl,
-      const std::vector<c10::Device>& devices) {
-    if (impl == nullptr) {
-      return {};
-    }
-    c10::DeviceIndex deviceCount = impl->deviceCount();
-    std::vector<bool> isDeviceUsed(deviceCount, false);
-    for (const c10::Device& device : devices) {
+  // We need devices to be sorted in order to use ensureIsSubsetOfDevices.
+  static std::vector<c10::Device> sortAndDeduplicateDevices(
+      const c10::impl::VirtualGuardImpl& impl,
+      std::vector<c10::Device> devices) {
```

Comment:
I might have ended up overengineering this function. I was just bothered that we were doing two heap allocations and throwing away the original vector when, in practice, we could often just use it as is.

I also wish we could just do `std::sort(...)` but we need to remove duplicates and that's tricky.

Reply:
This looks OK to me. Curious, will 1) sort devices + 2) two pointers/indices be a little shorter to implement?

Reply:
That's a good idea!